src/internal/agent/hooks/proxy.go

```diff
@@ -0,0 +1,224 @@
+
+			if strings.HasPrefix(req.URL.Path, noTransform) {
+				targetURL = noTransformTarget(zarfState.GitServer.Address, req.URL.Path)
+			} else {
+				// TODO: (@WSTARR) Remove hardcoded https, this doesn't come through on scheme, but we could check it from req.TLS (though we only serve https right now anyway)
+				transformedURL, err := git.TransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+				message.Debugf("%#v", err)
+
+				targetURL, err = url.Parse(transformedURL)
+				message.Debugf("%#v", err)
+			}
+		} else {
+			req.SetBasicAuth(zarfState.GitServer.PushUsername, zarfState.GitServer.PushPassword)
+
+			if strings.HasPrefix(req.URL.Path, noTransform) {
+				targetURL = noTransformTarget(zarfState.GitServer.Address, req.URL.Path)
+			} else {
+				transformedURL := ""
+
+				if strings.HasPrefix(req.UserAgent(), "pip") || strings.HasPrefix(req.UserAgent(), "twine") {
+					transformedURL = pipTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate git repos and package registries in the future
+				} else if strings.HasPrefix(req.UserAgent(), "npm") || strings.HasPrefix(req.UserAgent(), "yarn") {
+					transformedURL = npmTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate external git repos and package registries in the future
+					req.Header.Set("Authorization", "Bearer "+string(npmToken))
+				} else {
+					transformedURL = genTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate external git repos and package registries in the future
+				}
+
+				targetURL, err = url.Parse(transformedURL)
+				message.Debugf("%#v", err)
+			}
+		}
+
+		req.Host = targetURL.Host
+		req.URL = targetURL
+
+		message.Debugf("After Req %#v", req)
+		message.Debugf("After Req URL%#v", req.URL)
+	}
+
+	response := func(resp *http.Response) error {
+		message.Debugf("Before Resp %#v", resp)
+
+		// Handle redirection codes (3xx) by adding a marker to let Zarf know this has been redirected
+		if resp.StatusCode/100 == 3 {
+			message.Debugf("Before Resp Location %#v", resp.Header.Get("Location"))
+
+			locationURL, err := url.Parse(resp.Header.Get("Location"))
+			message.Debugf("%#v", err)
+			locationURL.Path = noTransform + locationURL.Path
+			locationURL.Host = resp.Request.Header.Get("X-Forwarded-Host")
+
+			resp.Header.Set("Location", locationURL.String())
+
+			message.Debugf("After Resp Location %#v", resp.Header.Get("Location"))
+		}
+
+		contentType := resp.Header.Get("Content-Type")
+		message.Debugf("%s", contentType)
+
+		// TODO: (@WSTARR) Refactor to be more concise/descriptive
```

Comment:
I think this block could use just a little bit of tidying up with error checking/handling but other than that I don't think it really needs that much of a refactor.

Reply:
Yeah it just needs to be made more readable really...

Reply:
I moved this to a helper function! (still could probably use better error handling still so not resolving yet)

Reply:
fixed error handling!